File: interpreter/src/main/java/cambio/tltea/interpreter/nodes/cause/ImplicationNode.java
package cambio.tltea.interpreter.nodes.cause;

import cambio.tltea.interpreter.nodes.StateChangeEvent;
import cambio.tltea.interpreter.nodes.StateChangeListener;
import cambio.tltea.interpreter.nodes.TriggerNotifier;

/**
 * @author Lion Wagner
 */
public class ImplicationNode extends InteractionNode<Boolean> implements StateChangeListener<Boolean> {

    private final InteractionNode<Boolean> cause;
    private final String consequence;
    private final TriggerNotifier notifier;

    public ImplicationNode(InteractionNode<Boolean> cause, String consequence, TriggerNotifier notifier) {
        this.cause = cause;
        this.consequence = consequence;
        this.notifier = notifier;

        cause.subscribe(this);
    }

    @Override
    public void onEvent(StateChangeEvent<Boolean> event) {
        if (event.getNewValue()) {
            notifier.trigger(consequence);
        }
    }

    /**
     * @return whether the cause expression is satisfied
     */
    @Override
    public Boolean getValue() {
        return cause.getValue();
    }
}

File: interpreter/src/main/java/cambio/tltea/interpreter/nodes/ImplicationNode.kt
package cambio.tltea.interpreter.nodes

import cambio.tltea.interpreter.nodes.consequence.ConsequenceNode
import cambio.tltea.parser.core.temporal.TemporalOperatorInfo

/**
 * @author Lion Wagner
 */
class ImplicationNode(
    val causeDescription: CauseDescription,
    val consequence: ConsequenceDescription,
    temporalContext: TemporalOperatorInfo,
    triggerManager: TriggerManager
) : ConsequenceNode(triggerManager, temporalContext), StateChangeListener<Boolean> {

    override fun activateConsequence() {
        //activate listeners to input of cause
        causeDescription.activateListeners()

        //activate self listening to changes
        causeDescription.causeChangePublisher.subscribe(this)
    }

    override fun deactivateConsequence() {
        causeDescription.deactivateListeners()
        causeDescription.causeChangePublisher.unsubscribe(this)
    }

    override fun onEvent(event: StateChangeEvent<Boolean>) {
        //activate consequence if the new value carried by the event is "true"
        if (event.newValue) {
            consequence.activateConsequence()
        }
        else {
            consequence.deactivateConsequence()
        }
    }

    /**
     * @return whether the cause expression is satisfied
     */
    val currentValue: Boolean
        get() = causeDescription.causeASTRoot.currentValue

}